core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -561,79 +560,62 @@ class GroupMetadataManager(val brokerId: Int,
-        }
-        catch {
-          case t: Throwable =>
-            error(s"Failed to write ${tombstones.size} tombstones for group ${group.groupId} to $appendPartition.", t)
-          // ignore and continue
-        }
+      val offsetsPartition = partitionFor(groupId)
+      getMessageFormatVersionAndTimestamp(offsetsPartition) match {
+        case Some((magicValue, timestamp)) =>
+          val partitionOpt = replicaManager.getPartition(Topic.GroupMetadataTopicName, offsetsPartition)
+          partitionOpt.foreach { partition =>
+            val appendPartition = TopicAndPartition(Topic.GroupMetadataTopicName, offsetsPartition)
+            val tombstones = expiredOffsets.map { case (topicPartition, offsetAndMetadata) =>
+              trace(s"Removing expired offset and metadata for $groupId, $topicPartition: $offsetAndMetadata")
+              val commitKey = GroupMetadataManager.offsetCommitKey(groupId, topicPartition.topic, topicPartition.partition)
+              new Message(bytes = null, key = commitKey, timestamp = timestamp, magicValue = magicValue)
+            }.toBuffer
+            trace(s"Marked ${expiredOffsets.size} offsets in $appendPartition for deletion.")
+
+            // We avoid writing the tombstone when the generationId is 0, since this group is only using
+            // Kafka for offset storage.
+            if (groupIsDead && groupMetadataCache.remove(groupId, group) && generation > 0) {
+              // Append the tombstone messages to the partition. It is okay if the replicas don't receive these (say,
+              // if we crash or leaders move) since the new leaders will still expire the consumers with heartbeat and
+              // retry removing this group.
+              tombstones += new Message(bytes = null, key = GroupMetadataManager.groupMetadataKey(group.groupId),
+                timestamp = timestamp, magicValue = magicValue)
+              trace(s"Marked group $groupId in $appendPartition for deletion")
+            }
 
-        group synchronized {
-          if (group.is(Dead)) {
-            groupMetadataCache.remove(group.groupId, group)
-            info("Group %s generation %s is dead and removed".format(group.groupId, group.generationId))
+            if (tombstones.nonEmpty) {
+              try {
+                // do not need to require acks since even if the tombstone is lost,
+                // it will be appended again in the next purge cycle
+                partition.appendMessagesToLeader(new ByteBufferMessageSet(config.offsetsTopicCompressionCodec, tombstones: _*))
+                offsetsRemoved += expiredOffsets.size
+                trace(s"Successfully appended tombstones to $appendPartition for expired offsets and/or metadata for group $groupId")
+              } catch {
+                case t: Throwable =>
+                  error(s"Failed to append tombstones to $appendPartition for expired offsets and/or metadata for group $groupId.", t)
```

Comment:
We used to report the number of tombstones, is there a reason why you removed that?